src/daemon/ca/status.rs

```diff
@@ -164,107 +245,225 @@ impl StatusStore {
     ) -> KrillResult<()> {
         let error_response = Self::error_to_error_res(error);
         self.update_ca_child_status(ca, child, |status| status.set_failure(user_agent, error_response))
-            .await
     }
 
     /// Marks a child as suspended. Note that it will be implicitly unsuspended whenever a new success or
     /// or failure is recorded for the child.
-    pub async fn set_child_suspended(&self, ca: &Handle, child: &ChildHandle) -> KrillResult<()> {
+    pub fn set_child_suspended(&self, ca: &Handle, child: &ChildHandle) -> KrillResult<()> {
         self.update_ca_child_status(ca, child, |status| status.set_suspended())
-            .await
-    }
-
-    /// Adds a child with default status values if the child is missing
-    pub async fn set_child_default_if_missing(&self, ca: &Handle, child: &ChildHandle) -> KrillResult<()> {
-        self.update_ca_child_status(ca, child, |_status| {}).await
     }
 
     /// Remove a CA from the saved status
     /// This should be called when the CA is removed from Krill, but note that if this is done for a CA which still exists
     /// a new empty default status will be re-generated when it is accessed for this CA.
-    pub async fn remove_ca(&self, ca: &Handle) -> KrillResult<()> {
-        let lock = self.locks.ca(ca).await;
-        let _ = lock.write().await;
+    pub fn remove_ca(&self, ca: &Handle) -> KrillResult<()> {
+        self.cache.write().unwrap().remove(ca);
 
         let scope = ca.as_str();
-
-        self.store.drop_scope(scope)?; // will fail in case of I/O errors only
+        self.store.drop_scope(scope)?; // will only fail if scope is present and cannot be removed
 
         Ok(())
     }
 
     /// Removes a child for the given CA.
-    pub async fn remove_child(&self, ca: &Handle, child: &ChildHandle) -> KrillResult<()> {
-        self.update_ca_status(ca, |status| {
-            status.children.remove(child);
-        })
-        .await
+    pub fn remove_child(&self, ca: &Handle, child: &ChildHandle) -> KrillResult<()> {
+        let mut cache = self.cache.write().unwrap();
+
+        if let Some(mut ca_status) = cache.get_mut(ca) {
+            let ca_status = Arc::make_mut(&mut ca_status);
```

Comment:
The documentation for `make_mut()` says "If there are other Arc pointers to the same allocation, then make_mut will clone the inner value to a new allocation to ensure unique ownership. This is also referred to as clone-on-write.". You would then be calling `.children.remove(child)` on a clone of the `children` `HashMap`, i.e. not actually removing anything from the _actual_ children HashMap, am I understanding that right and is that really what is wanted?

Reply:
doh! good point.. I had it reversed somehow..

That said, this triggered me to rethink it all. Read operations here are infrequent - where we need a copy of the data that is -while write operations are frequent. So not using an Arc at all actually simplifies the code quite a bit and makes the updates more efficient.